### PR TITLE
=clu #13875 Exclude unreachability observations from downed

### PR DESCRIPTION
**Pull request by patriknw**
- Skip observations from downed node (quarantined is marked down immediately)
  in convergence check
- Skip observations from downed node when picking "reachable" targets for gossip.
- This also means that we must accept gossip with own node marked as unreachable,
  but that should not be spread to the external membership events.

The SurviveNetworkInstabilitySpec failed a lot before this patch. I have verified that it does not fail any more by letting jenkins run this over the weekend.
744 successful runs of SurviveNetworkInstabilitySpec
90  successful runs of all tests in akka-cluster.

This PR also contains a cherry-pick of forgotten forward port of adjustments to SurviveNetworkInstabilitySpec.

### CODE REVIEW

**Comment by patriknw** [comment]
Refs #13875
Refs #15439

**Comment by hepin1989** [comment]
will this related to https://github.com/akka/akka/issues/16224 ?

**Comment by patriknw** [comment]
> does this related to 16224
I don't think so. I'm stuck on 16224, in the sense that I don't know what the problem is apart from the thing I have already mentioned. Please continue discussion there.

File: akka-cluster/src/test/scala/akka/cluster/ClusterDomainEventPublisherSpec.scala [hunk 1]
```diff
@@ -19,8 +19,15 @@ import akka.testkit.ImplicitSender
...
+object ClusterDomainEventPublisherSpec {
+  val config = """
+    akka.actor.provider = "akka.cluster.ClusterActorRefProvider"
+    akka.remote.netty.tcp.port = 0
+    """
+}
+
 @org.junit.runner.RunWith(classOf[org.scalatest.junit.JUnitRunner])
-class ClusterDomainEventPublisherSpec extends AkkaSpec
+class ClusterDomainEventPublisherSpec extends AkkaSpec(ClusterDomainEventPublisherSpec.config)
```
Comment (rkuhn) [inline akka-cluster/src/test/scala/akka/cluster/ClusterDomainEventPublisherSpec.scala]: I’m slightly confused: did this test suite not work before?

**Comment by rkuhn** [comment]
LGTM, nice work!

File: akka-cluster/src/test/scala/akka/cluster/ClusterDomainEventPublisherSpec.scala (hunk 1 above)
Comment (patriknw) [inline akka-cluster/src/test/scala/akka/cluster/ClusterDomainEventPublisherSpec.scala]: yes, the change is that the ClusterDomainEventPublisher actor now make use of `selfUniqueAddress` from the `Cluster` extension, and therefore this test must use the CARP

Comment (rkuhn) [inline akka-cluster/src/test/scala/akka/cluster/ClusterDomainEventPublisherSpec.scala]: Ah, right, I didn’t connect the dots.

**Comment by patriknw** [comment]
Can I merge this so that we get rid of the jenkins failures, or will someone more review it?

**Comment by bantonsson** [comment]
Sorry, didn't finish it Tuesday.  I'll do that now.

File: akka-cluster/src/main/scala/akka/cluster/Gossip.scala [hunk 2]
```diff
@@ -159,19 +159,26 @@ private[cluster] final case class Gossip(
...
+    val unreachable = reachabilityExcludingDownedObservers.allUnreachableOrTerminated.collect {
+      case node if (node != selfUniqueAddress) ⇒ member(node)
+    }
     unreachable.forall(m ⇒ Gossip.convergenceSkipUnreachableWithMemberStatus(m.status)) &&
       !members.exists(m ⇒ Gossip.convergenceMemberStatus(m.status) && !seenByNode(m.uniqueAddress))
   }
 
+  lazy val reachabilityExcludingDownedObservers: Reachability = {
+    val downed = members.collect { case m if m.status == Down ⇒ m }
+    if (downed.isEmpty) overview.reachability else overview.reachability.removeObservers(downed.map(_.uniqueAddress))
```
Comment (bantonsson) [inline akka-cluster/src/main/scala/akka/cluster/Gossip.scala]: Should this optimization be in `removeObservers`? I mean do nothing on an empty `Set`.

**Comment by bantonsson** [comment]
LGTM

File: akka-cluster/src/main/scala/akka/cluster/Gossip.scala (hunk 2 above)
Comment (patriknw) [inline akka-cluster/src/main/scala/akka/cluster/Gossip.scala]: good point,
`downed.map(_.uniqueAddress)` allocates something also when downed is empty, but I guess that is neglectable

Comment (bantonsson) [inline akka-cluster/src/main/scala/akka/cluster/Gossip.scala]: Yes, I think that object allocation is negligible.

**Comment by rkuhn** [comment]
Merge away!

File: akka-cluster/src/main/scala/akka/cluster/Gossip.scala (hunk 2 above)
Comment (patriknw) [inline akka-cluster/src/main/scala/akka/cluster/Gossip.scala]: ah, that is how it is spelled :)
I'll fix that and merge.

Comment (rkuhn) [inline akka-cluster/src/main/scala/akka/cluster/Gossip.scala]: I’d even say “spelt” :-)

Comment (patriknw) [inline akka-cluster/src/main/scala/akka/cluster/Gossip.scala]: always learning from these reviews :-)

**Pull request merged**